src/pyhf/pdf.py

```diff
@@ -531,16 +538,20 @@ def logpdf(self, pars, data):
         try:
             tensorlib, _ = get_backend()
             pars, data = tensorlib.astensor(pars), tensorlib.astensor(data)
-            cut = tensorlib.shape(data)[0] - len(self.config.auxdata)
-            actual_data, aux_data = data[:cut], data[cut:]
 
-            mainpdf = self.mainlogpdf(actual_data, pars)
-            constraint = self.constraint_logpdf(aux_data, pars)
+            actual_data = self.main_model._dataprojection(data)
+            aux_data = self.constraint_model._dataprojection(data)
+
+            mainpdf = self.main_model.logpdf(actual_data, pars)
```

Comment:
thinking about whether `main_model.logpdf` should call `_dataprojection` or if `pyhf/pdf.py` should. You did originally call it as `_dataprojection` which implies you only ever expected it to be an internal call, but you're using it as a non-internal call here. It does certainly look cleaner to do

```
mainpdf = self.main_model.logpdf(data, pars)
constraintpdf = self.constraint_model.logpdf(data, pars)
```

Additionally, please call it `constraintpdf` and not `constraint`.

Reply:
done